### PR TITLE
Fix various NREs. Calculate vessel orientation without using the NavBall, which doesn't exist in IVA mode.

### PR DESCRIPTION
**Pull request by BrettRyland**
This fixes the massive performance penalty (see profiling pics) and failure to update the vessel orientation (which likely causes many other issues) in IVA mode caused by a call to `FindObjectOfType<NavBall>()` (which doesn't exist in IVA mode) in `GetNavball` (which is being called every `FixedUpdate` since it doesn't find it). It replaces the `NavBall` code with explicit calculations of the heading, pitch and roll (which I've checked to be consistent with the values from NavBall to around single precision when not in IVA mode).

FAR - FindObjectOfType in GetNavBall in IVA view
FAR-FindObjectsOfType

Additionally, this fixes various `NullReferenceException`s related to EVA kerbals' parachutes and parts that have been destroyed.

### CODE REVIEW

File: FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs [hunk 1]
```diff
@@ -443,21 +443,21 @@ public void AssumeDragCubePosition(string cubeName)
                 case "STOWED":
                 {
                     parachute.gameObject.SetActive(false);
-                    cap.gameObject.SetActive(true);
+                    if (cap != null) cap.gameObject.SetActive(true);
```
Comment (dkavolis) [inline FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs:446]: I don't think any of the `cap != null` checks are needed here, `cap == null` shows a bad installation that should be fixed instead

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs [hunk 2]
```diff
@@ -547,6 +547,8 @@ public void CalculateSurfaceFunctions()
             float mass = 0;
             foreach (Part p in VesselPartList)
             {
+                if (p == null)
+                    continue;
```
Comment (dkavolis) [inline FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: Have you ever encountered `null`s here? `p` should never be null unless KSP returns such part lists

File: FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs [hunk 3]
```diff
@@ -215,17 +215,43 @@ private void CalculateVesselOrientation(Vector3d velVectorNorm)
...
+            // vesselInfo.rollAngle =
+            //     vesselRot.eulerAngles.z > 180 ? 360 - vesselRot.eulerAngles.z : -vesselRot.eulerAngles.z;
+
+            // Gives the same numbers as the above, but also works in IVA mode which doesn't have a NavBall instance.
+            var localUp = (_vessel.transform.position - _vessel.mainBody.transform.position).normalized;
+            var east = Vector3.Cross(localUp, _vessel.mainBody.RotationAxis).normalized;
+            var north = Vector3.Cross(east, localUp).normalized;
+            vesselInfo.headingAngle = (SignedAngle(north, Vector3.ProjectOnPlane(up, localUp).normalized, localUp) + 360f) % 360f;
+            vesselInfo.pitchAngle = 90f - SignedAngle(localUp, up, Vector3.Cross(localUp, up).normalized);
+            vesselInfo.rollAngle = SignedAngle(Vector3.Cross(localUp, up).normalized, right, -up);
```
Comment (dkavolis) [inline FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs:236]: Why not use `Quaternion` methods here?

**Review by dkavolis — COMMENTED** [review]
Thanks for the PR. I've left some comments before merging.

File: FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs (hunk 1 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs:446]: For parachutes on EVA kerbals in command seats, the `cap = part.FindModelTransform(capName);` on line 880 of RealChuteFAR.cs returns `null`. This is possibly due to a local MM patch that renames `ModuleEvaChute` as `RealChuteFAR` (which is possibly not the best). In any case, if the `cap` transform isn't found for some reason, then the `if (cap != null)` checks need to be made to avoid NREs.

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 2 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: Yes, quite a lot of them in BDA tournaments with proc wings. This is likely due to the ordering of events being resolved since both `UpdateShipPartsList` and `CalculateSurfaceFunctions` are triggered by `OnVesselPartsChange`. Perhaps this should be calling `GetShipPartList` prior to iterating through the list instead of just skipping the parts if they're destroyed?

File: FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs (hunk 3 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs:236]: I couldn't find the correct quaternion to use and you're immediately converting them to Euler angles anyway.

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 2 above)
Comment (dkavolis) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: Really, `UpdateShipPartsList` should be called in `ForceOnVesselPartsChange` before invoking `OnVesselPartsChange` to not rely on the invocation order.

File: FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs (hunk 1 above)
Comment (dkavolis) [reply FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs:446]: In that case, `cap?.gameObject` would be more appropriate and consice as `cap` is never a valid Unity object in the first place.

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 2 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: But `ForceOnVesselPartsChange` is called from `ForceLegacyAeroUpdates`, which is only called in `FixedUpdate` in FARVesselAero.cs, so it wouldn't be being updated in time for the event handling.

File: FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs (hunk 1 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/RealChuteLite/RealChuteFAR.cs:446]: Yes, that'd be better.

**Comment by BrettRyland** [comment]
Found another NRE, this time caused by
```
Dictionary<Part, VoxelCrossSection.SideAreaValues> partSideAreas =
    crossSections[m].partSideAreaValues;
partSideAreas.Clear();
```
```
[[LOG 17:01:04.321] B9.Aero.Wing.Procedural.TypeB Exploded!! - blast awesomeness: 0.1
[LOG 17:01:04.343] [FAR v0.16.0.5]: Updating vessel voxel for Reverse Mosquito Mk Vd_1 Debris
[ERR 17:01:04.348] [FAR v0.16.0.5]: Voxel Volume was infinity; ending voxelization
        UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])
        FerramAerospaceResearch.FARLogHandler:Log(LogLevel, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:40)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:81)
[LOG 17:01:04.349] [FAR v0.16.0.5]: Voxel Element CrossSection Area: 0
EXC 17:01:04.350] [FAR v0.16.0.5]: Logged exception:
        UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:77)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
[EXC 17:01:04.352] NullReferenceException: Object reference not set to an instance of an object
        FerramAerospaceResearch.FARPartGeometry.VehicleVoxel.CrossSectionData (FerramAerospaceResearch.FARPartGeometry.VoxelCrossSection[] crossSections, UnityEngine.Vector3 orientationVector, System.Int32& frontIndex, System.Int32& backIndex, System.Double& sectionThickness, System.Double& maxCrossSectionArea) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARPartGeometry/VehicleVoxel.cs:461)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CalculateVesselAeroProperties () (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:1018)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CreateVoxel (Vessel vessel) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:403)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:78)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
```

File: FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs (hunk 3 above)
Comment (dkavolis) [reply FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs:236]: Maybe something like this
```csharp
var offsetAngles = new Vector3(_vessel.isEVA ? 90f : 0f, 0f, 0f);
var offset = Quaternion.Euler(offsetAngles);
Quaternion attitude = offset * Quaternion.Inverse(refTransform.rotation);
Transform transform = _vessel.mainBody.transform;
Vector3 position = transform.position;
Vector3 refPosition = refTransform.position;
Vector3 localUp = (refPosition - position).normalized;
var vesselRot = Quaternion.Inverse(attitude *
                                   Quaternion.LookRotation(Vector3.ProjectOnPlane(position +
                                                                          (Vector3d)transform.up *
                                                                          _vessel.mainBody.Radius -
                                                                          refPosition,
                                                                      localUp)
                                                                  .normalized,
                                                           localUp));
```

**Comment by dkavolis** [comment]
I don't have any ideas why the last NRE is there. `crossSections` is always created by
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/FARPartGeometry/VehicleVoxel.cs#L95-L106
used in https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs#L398

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 2 above)
Comment (dkavolis) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: `ForceOnVesselPartsChange` also invokes `CalculateSurfaceFunctions`. Can you try calling `UpdateShipPartsList` in `ForceOnVesselPartsChange` before the delegate to see if it changes anything?

File: FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs (hunk 3 above)
Comment (dkavolis) [reply FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs:236]: Please do.

Comment (BrettRyland) [reply FerramAerospaceResearch/FARGUI/FARFlightGUI/PhysicsCalcs.cs:236]: It looks like the quaternion you give above doesn't give the correct angles. Here's the FAR flight data GUI extended with the results of the orientation calculations at the bottom and the KER vessel GUI for confirmation of the orientation (KER seems to be a bit inaccurate). The `err` values are the difference between the original code with `NavBall` and the two approaches above.
Screenshot_20220502_193121

**Comment by BrettRyland** [comment]
Yet another exception appeared:
```
[EXC 19:41:43.748] [FAR v0.16.0.5]: Logged exception:
        UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:77)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
[EXC 19:41:43.751] KeyNotFoundException: The given key was not present in the dictionary.
        System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) (at <9577ac7a62ef43179789031239ba8798>:0)
        FerramAerospaceResearch.FARAeroComponents.FARAeroSection.UpdateAeroSection (System.Single potentialFlowNormalForce, System.Single viscCrossflowDrag, System.Single diameter, System.Single flatnessRatio, System.Single hypersonicMomentForward, System.Single hypersonicMomentBackward, UnityEngine.Vector3 centroidWorldSpace, UnityEngine.Vector3 xRefVectorWorldSpace, UnityEngine.Vector3 nRefVectorWorldSpace, UnityEngine.Matrix4x4 vesselToWorldMatrix, UnityEngine.Vector3 vehicleMainAxis, System.Collections.Generic.List`1[T] moduleList, System.Collections.Generic.List`1[T] dragFactor, System.Collections.Generic.Dictionary`2[TKey,TValue] partWorldToLocalMatrixDict) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:164)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CalculateVesselAeroProperties () (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:1485)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CreateVoxel (Vessel vessel) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:403)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:78)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
```

I think it needs a
```
if (!partWorldToLocalMatrixDict.ContainsKey(data.aeroModule.part))
                    continue;
```
between lines 163 and 164 of FARAeroSection.cs like in the block above it (lines 146-153).

**Comment by dkavolis** [comment]
Good catch, but instead of using `Contains` use `TryGetValue` to avoid double lookup, and do the same for the block above.

**Comment by BrettRyland** [comment]
> I don't have any ideas why the last NRE is there. `crossSections` is always created by
> 
> https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/FARPartGeometry/VehicleVoxel.cs#L95-L106

Ah, I think I see what's going on here. `EmptyCrossSectionArray` is called from `CreateVoxel` if `_vehicleCrossSection.Length < _voxel.MaxArrayLength`, but `VehicleVoxel.CreateNewVoxel` returns early due to "Voxel Volume was infinity; ending voxelization" so `_voxel.MaxArrayLength` is 0 and the `if` condition is never satisfied. 
However, `_vehicleCrossSection` is created with size 1 initially on line 74 of VehicleAeroDynamics.cs, so if it doesn't get updated before the first use, the 0th index will be `null` and `sectionCount` from
```
int sectionCount = yCellLength + (int)(xCellLength * x / y + 1) + (int)(zCellLength * z / y + 1);
sectionCount = Math.Min(sectionCount, crossSections.Length);
```
on lines 428-429 of VehicleVoxel.cs allows this first index to be accessed.

Line 398 of VehicleAerodynamics.cs seems to be the only place where `_vehicleCrossSection` is assigned. There probably needs to be an `else` condition to that `if` so that it's not reusing old values (or a `null` value on the first use) in the `crossSections` array on line 459-461.

**Comment by dkavolis** [comment]
Changing https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs#L74 to `new VoxelCrossSection[0]` should work, it's the only place where the contained dict is not created.

**Comment by BrettRyland** [comment]
> Changing
> 
> https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs#L74
> 
> to `Array.Empty<VoxelCrossSection>()` should work, it's the only place where the contained dict is not created.

That would avoid the `null`, but what about the case when `_voxel`is smaller than `_vehicleCrossSection` in
```
_voxel = VehicleVoxel.CreateNewVoxel(_currentGeoModules, _voxelCount, vessel: vessel);
if (_vehicleCrossSection.Length < _voxel.MaxArrayLength)
    _vehicleCrossSection = _voxel.EmptyCrossSectionArray;
```
? `CalculateVesselAeroProperties` would then be running on an old `_vehicleCrossSection`.

**Comment by dkavolis** [comment]
I think it's still better than not using anything, at least the old value was valid whereas the new one would blow up the physics due to infinities.

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 2 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: I haven't had any exceptions from this since adding the `UpdateShipPartsList` in `ForceOnVesselPartsChange`, so I guess that makes the `p == null` check unnecessary.

Comment (dkavolis) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: Yeah, if rather not have have Unity Object comparisons to null if possible

Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: Ah, no, wait, I had the debug log level wrong in the `FARLogger` and it just wasn't printing the debug message I was looking for. `VesselPartList` does still contain `null` entries.

Comment (dkavolis) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: In this case it might be best to remove all null parts in `UpdateShipPartsList` instead of doing null checks in multiple places

Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: I still think it's being caused by the ordering of event handling. I'm going to test adding a call to `UpdateShipPartsList` right before that loop to see if that helps first.

Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:551]: That seemed to have worked. The last 11 rounds of my tournament ran without any exceptions.

**Comment by BrettRyland** [comment]
OK, I think I've fixed all the above points now. The only thing is whether you want to try again to get a quaternion approach to the vessel orientation in PhysicsCalcs.cs or just use the vector algebra.

In any case, I'm going to head to bed now. I'll run some more tests tomorrow (SuicidalInsanity is running a new round of BAD-T using FAR, which is the reason I started looking at this in the first place).

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs [hunk 4]
```diff
@@ -545,6 +545,7 @@ public void CalculateSurfaceFunctions()
 
             Vector3 CoM = Vector3.zero;
             float mass = 0;
+            UpdateShipPartsList();
```
Comment (dkavolis) [inline FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:548]: `FARWingAerodynamicModel` also uses the part list. This should be in
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/LEGACYferram4/FARPartModule.cs#L69-L72
before `OnVesselPartsChange` is invoked and
https://github.com/dkavolis/Ferram-Aerospace-Research/blob/95e127ae140b4be9699da8783d24dd8db726d753/FerramAerospaceResearch/LEGACYferram4/FARPartModule.cs#L81
can be removed

**Review by dkavolis — CHANGES_REQUESTED** [review]

File: FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs (hunk 4 above)
Comment (BrettRyland) [reply FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs:548]: I've run some more tests to see exactly when the `null` parts are showing up in relation to the `OnVesselPartsChange` events and it turns out that they're actually occurring from the
```
if (justStarted)
    CalculateSurfaceFunctions();
```
call on lines 469 - 470 of FARControllableSurface.cs instead, which can occur just after a plane hits the terrain and the section of the plane containing the part becomes a new vessel.
Adding a `UpdateShipPartsList` to this `if` block seems to fix the issue.

(`UpdateShipPartsList` should still be called before `OnVesselPartsChange` is invoked though to ensure that the list is up-to-date for `CalculateSurfaceFunctions` and `UpdateThisWingInteractions`.)

**Comment by BrettRyland** [comment]
Another exception came up:
```
[LOG 12:24:36.552] [FAR v0.16.0.5]: Updating vessel voxel for CA-15 Kangaroo Debris
[LOG 12:24:36.568] miniFuselage Exploded!! - blast awesomeness: 0.5
[LOG 12:24:36.631] [FAR v0.16.0.5]: Voxel Element CrossSection Area: 0.0149669817358196
[LOG 12:24:36.631] [FAR v0.16.0.5]: Std dev for smoothing: 3 voxel total vol: 0.366210960783051 filled vol: 1.36691105828926
[EXC 12:24:36.633] [FAR v0.16.0.5]: Logged exception:
  UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])  
  ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])  
  FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:77)  
  FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
[EXC 12:24:36.636] IndexOutOfRangeException: Index was outside the bounds of the array.
  FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CalculateSonicPressure (FerramAerospaceResearch.FARPartGeometry.VoxelCrossSection[] vehicleCrossSection, System.Int32 front, System.Int32 back, System.Double sectionThickness) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:1695)  
  FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CalculateVesselAeroProperties () (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:1092)  
  FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CreateVoxel (Vessel vessel) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:403)  
  UnityEngine.DebugLogHandler:LogException(Exception, Object)  
  ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)  
  FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:78)  
  FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
[WRN 12:24:36.696] PlayOneShot was called with a null AudioClip.
```

This appears to stem from using a forward finite difference to approximate the first derivative on line 1695 of VehicleAerodynamics.cs, where the `i + 1` is going beyond the end of the `vehicleCrossSection` array if `back` remains at `crossSections.Length - 1` from line 410 of VehicleVoxel.cs.
You should probably be using a backward finite difference (using `i - 1` and `i` instead of `i` and `i + 1`) for the `back` index. Similarly for line 1654.

**Comment by BrettRyland** [comment]
And another exception:
```
[EXC 14:54:23.213] [FAR v0.16.0.5]: Logged exception:
        UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogFormat(LogType, Object, String, Object[])
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:77)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
[EXC 14:54:23.231] ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
        System.ThrowHelper.ThrowArgumentOutOfRangeException (System.ExceptionArgument argument, System.ExceptionResource resource) (at <9577ac7a62ef43179789031239ba8798>:0)
        System.ThrowHelper.ThrowArgumentOutOfRangeException () (at <9577ac7a62ef43179789031239ba8798>:0)
        FerramAerospaceResearch.FARAeroComponents.FARAeroSection.MergeAeroSection (FerramAerospaceResearch.FARAeroComponents.FARAeroSection otherSection) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/FARAeroSection.cs:283)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CalculateVesselAeroProperties () (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:1503)
        FerramAerospaceResearch.FARAeroComponents.VehicleAerodynamics.CreateVoxel (Vessel vessel) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARAeroComponents/VehicleAerodynamics.cs:403)
        UnityEngine.DebugLogHandler:LogException(Exception, Object)
        ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
        FerramAerospaceResearch.FARLogHandler:LogException(Exception, String, Object) (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch.Base/Utils/FARLogHandler.cs:78)
        FerramAerospaceResearch.FARThreading.ThreadSafeDebugLogger:Update() (at /home/brett/github/Ferram-Aerospace-Research/FerramAerospaceResearch/FARThreading/ThreadSafeDebugLogger.cs:79)
```
I'm not sure how this one occurs. The only potentially questionable indexing is `partData[index]` where `index` is found from `handledAeroModulesIndexDict.TryGetValue(tmpOtherData.aeroModule, out int index)`, but the only way I can see that causing a bad index is if `partData` and `handledAeroModulesIndexDict` aren't in sync.

**Comment by dkavolis** [comment]
> This appears to stem from using a forward finite difference to approximate the first derivative on line 1695 of VehicleAerodynamics.cs, where the `i + 1` is going beyond the end of the `vehicleCrossSection` array if `back` remains at `crossSections.Length - 1` from line 410 of VehicleVoxel.cs.
> You should probably be using a backward finite difference (using `i - 1` and `i` instead of `i` and `i + 1`) for the `back` index. Similarly for line 1654.

You're right, even the two branches are identical.

> I'm not sure how this one occurs. The only potentially questionable indexing is `partData[index]` where `index` is found from `handledAeroModulesIndexDict.TryGetValue(tmpOtherData.aeroModule, out int index)`, but the only way I can see that causing a bad index is if `partData` and `handledAeroModulesIndexDict` aren't in sync.

They look to be updated in the same places so they should always be synced unless there was an exception between updating the two.

**Review by dkavolis — APPROVED** [review]

**Comment by dkavolis** [comment]
Are you going to try and fix the last NRE? If not, I'll merge the PR.

**Comment by BrettRyland** [comment]
> Are you going to try and fix the last NRE? If not, I'll merge the PR.

Probably not. I've only seen it once, so it must be fairly rare and I don't really see why it occurred.
I'll make a new PR if I find other things to fix. :+1:

**Comment by dkavolis** [comment]
Thanks!
